From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Wed, 8 Mar 2023 11:25:22 -0800
Subject: [PATCH] visibility changes


diff --git a/src/main/java/net/minecraft/stats/StatType.java b/src/main/java/net/minecraft/stats/StatType.java
index e99ce87fcd3a2f5edaef7b8996fb875ae61a88c1..1ffdd966af6585d5d9f5cead5c72f54b22a8fa4c 100644
--- a/src/main/java/net/minecraft/stats/StatType.java
+++ b/src/main/java/net/minecraft/stats/StatType.java
@@ -10,7 +10,7 @@ import net.minecraft.network.chat.Component;
 
 public class StatType<T> implements Iterable<Stat<T>> {
     private final Registry<T> registry;
-    private final Map<T, Stat<T>> map = new IdentityHashMap<>();
+    public final Map<T, Stat<T>> map = new IdentityHashMap<>(); // MachinedPaper
     @Nullable
     private Component displayName;
 
diff --git a/src/main/java/net/minecraft/world/item/alchemy/PotionBrewing.java b/src/main/java/net/minecraft/world/item/alchemy/PotionBrewing.java
index 12d9556a11ac4ef2e7a62fcd2686d868904bc010..cd14e2202738ee11eb636d49965c3431bd2f1137 100644
--- a/src/main/java/net/minecraft/world/item/alchemy/PotionBrewing.java
+++ b/src/main/java/net/minecraft/world/item/alchemy/PotionBrewing.java
@@ -12,7 +12,7 @@ import net.minecraft.world.item.crafting.Ingredient;
 
 public class PotionBrewing {
     public static final int BREWING_TIME_SECONDS = 20;
-    private static final List<PotionBrewing.Mix<Potion>> POTION_MIXES = Lists.newArrayList();
+    public static final List<PotionBrewing.Mix<Potion>> POTION_MIXES = Lists.newArrayList(); // MachinedPaper
     private static final List<PotionBrewing.Mix<Item>> CONTAINER_MIXES = Lists.newArrayList();
     private static final it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap<org.bukkit.NamespacedKey, io.papermc.paper.potion.PaperPotionMix> CUSTOM_MIXES = new it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap<>(); // Paper
     private static final List<Ingredient> ALLOWED_CONTAINERS = Lists.newArrayList();
@@ -273,10 +273,10 @@ public class PotionBrewing {
         POTION_MIXES.add(new PotionBrewing.Mix<>(input, Ingredient.of(item), output));
     }
 
-    static class Mix<T> {
-        final T from;
+    public static class Mix<T> { // MachinedPaper
+        public final T from; // MachinedPaper
         final Ingredient ingredient;
-        final T to;
+        public final T to; // MachinedPaper
 
         public Mix(T input, Ingredient ingredient, T output) {
             this.from = input;
